File: service_backend/app/models/team.py
"""Team + TeamMember models (plan 28, roadmap A8, D-A8-1).

A core `public.teams` grouping of tenant users, platform-wide, next to roles -
any future Service (omnichannel first) assigns work to a team instead of only
a single user. Tenant-scoped like every core entity; name is unique per tenant
case-insensitively (a Postgres functional index in the migration + a service-
level check so the sqlite test path behaves identically).

`TeamMember.tenant_id` is ALWAYS set from the owning Team at insert time (the
service layer does this explicitly, never trusting the payload) - the same
house rule as `user_roles.tenant_id` deriving from the owning role (BL-015).
"""
import uuid

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    ForeignKey,
    Integer,
    String,
    UniqueConstraint,
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base
from app.models.tenant import DEFAULT_TENANT_ID
from app.models.utc_datetime import UTCDateTime

TEAM_MEMBER_ROLES = ("member", "lead")


class Team(Base):
    __tablename__ = "teams"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    tenant_id = Column(
        String,
        ForeignKey("tenants.id"),
        nullable=False,
        index=True,
        default=DEFAULT_TENANT_ID,
    )
    name = Column(String, nullable=False)
    description = Column(String, nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)
    sort_order = Column(Integer, default=0, nullable=False)

    created_at = Column(UTCDateTime(), server_default=func.now(), nullable=False)
    updated_at = Column(
        UTCDateTime(), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    # Replace-set membership (create/update `members`) - delete-orphan so
    # reassigning `team.members = [...]` clears removed rows in one flush.
    members = relationship(
        "TeamMember",
        cascade="all, delete-orphan",
        order_by="TeamMember.created_at",
        lazy="selectin",
    )


class TeamMember(Base):
    __tablename__ = "team_members"
    __table_args__ = (
        UniqueConstraint("team_id", "user_id", name="uq_team_members_team_user"),
        CheckConstraint("role IN ('member','lead')", name="ck_team_members_role"),
    )

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    team_id = Column(
        String, ForeignKey("teams.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # Derived from the owning Team at insert (never from client payload) -
    # carried for tenant-scoped queries without a join (BL-015 pattern).
    tenant_id = Column(
        String,
        ForeignKey("tenants.id"),
        nullable=False,
        index=True,
        default=DEFAULT_TENANT_ID,
    )
    user_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    role = Column(String, nullable=False, default="member")

    created_at = Column(UTCDateTime(), server_default=func.now(), nullable=False)

    # Read-side convenience (name/email for TeamMemberRef) - joined so listing
    # a team's members costs no N+1.
    user = relationship("User", lazy="joined")

File: service_frontend/types/team.ts
/**
 * Team domain types (plan 28, roadmap A8). Mirrors the backend contract
 * (`documentation/plans/sprint-4/28-teams-core-and-omnichannel-assignment.md`
 * §5.1) a core `public.teams` grouping of tenant users, platform-wide, so any
 * Service (omnichannel first) can assign work to a team instead of only a
 * single user. `team-service.ts` binds to the real backend since S5.
 */

/** A team member's role - a label only in this slice (D-A8-17): no extra
 * permission, no assignment priority (B1 owns access levels). */
export type TeamMemberRole = 'member' | 'lead';

/** One member row on a team (`TeamMemberRef` in the backend contract). */
export interface TeamMemberRef {
  userId: string;
  name: string;
  email: string;
  role: TeamMemberRole;
}

/** A core team - tenant-scoped, shared by every Service. */
export interface Team {
  id: string;
  name: string;
  description: string | null;
  isActive: boolean;
  sortOrder: number;
  members: TeamMemberRef[];
  memberCount: number;
  createdAt: string; // ISO
  updatedAt: string; // ISO
}

/** One entry in the create/update `members` payload. */
export interface TeamMemberInput {
  userId: string;
  role: TeamMemberRole;
}

/** `GET /teams/mine` member shape (review round 1, finding 7) - NO `email`.
 * That route is authenticated-only (no `teams.read`), so it must never leak
 * teammates' email addresses to every signed-in user. */
export interface MyTeamMemberRef {
  userId: string;
  name: string;
  role: TeamMemberRole;
}

/** Trimmed `GET /teams/mine` team shape - no `description`/`sortOrder`/
 * timestamps and no member emails. */
export interface MyTeam {
  id: string;
  name: string;
  isActive: boolean;
  memberCount: number;
  members: MyTeamMemberRef[];
}

export interface CreateTeamInput {
  name: string;
  description?: string | null;
  isActive?: boolean;
  sortOrder?: number;
  members: TeamMemberInput[];
}

export interface UpdateTeamInput {
  name?: string;
  description?: string | null;
  isActive?: boolean;
  sortOrder?: number;
  members?: TeamMemberInput[];
}
